File: server/core/Player.js
var Cemetery = require("./Cemetery");
var Deck = require("./Deck");
var Field = require("./Field");
var Hand = require("./Hand");

var Player = function(deckList){
	this.hand = new Hand();
	this.field = new Field();
	this.cemetery = new Cemetery();
	this.canAttack = true;
	this.deck = new Deck(deckList);
	this.deck.shuffle();
	this.canDraw = 7;
	for(var i = 0; i < 7; i++){this.draw();}
};

Player.prototype.draw = function(){
	this.hand.add(this.deck.draw());
};

Player.prototype.draws = function(num){
	for(var i = 0; i < num; i++){
		this.hand.add(this.deck.draw());
	}
};

Player.prototype.firstStep = function(){
	var card = this.deck.firstStep();
	this.cemetery.add(card);
	return [card.number, card.mark];
};

Player.prototype.charge = function(){
	this.field.chargeAll();
	this.field.canAttackAll();
};

Player.prototype.cleanUp = function(length){
	if(this.hand.length > 7){
		this.cemetery.add(this.hand.get(length));
	}
};

Player.prototype.damage = function(){
	this.cemetery.add(this.deck.damage());
};

Player.prototype.damages = function(num){
	for(var i = 0; i < num; i++){
		this.damage();
	}
};

Player.prototype.setBarrier = function(length){
	this.damage();
	this.field.setBarrier(this.hand.get(length));
};

Player.prototype.summonSoldier = function(handLength, barrierLength){
	this.damage();
	this.field.drive("barrier", barrierLength);
	this.field.summon(this.hand.get(handLength));
};

Player.prototype.summonAce = function(handLength){
	this.damage();
	this.field.summon(this.hand.get(handLength));
};

Player.prototype.summonHero = function(handLength, barrierLength_1, barrierLength_2){
	this.damage();
	this.field.drive("barrier", barrierLength_1);
	this.field.drive("barrier", barrierLength_2);
	this.field.summon(this.hand.get(handLength));
};

Player.prototype.summonMagician = function(handLength, costHandLength, barrierLength){
	this.field.drive("barrier", barrierLength);
	if(handLength > costHandLength){
		this.field.summon(this.hand.get(handLength));
		this.cemetery.add(this.hand.get(costHandLength));
	}else{
		this.cemetery.add(this.hand.get(costHandLength));
		this.field.summon(this.hand.get(handLength));
	}
};

Player.prototype.declarationAttack = function(soldierLength){
	var attaker = [];
	for(var i = 0; i < soldierLength.length; i++){
		if(this.field.soldier[soldierLength[i]].canAttack){
			attaker.push(this.field.soldier[soldierLength[i]]);
			this.field.drive("soldier", soldierLength[i]);
		}
	}
	return attaker;
};

Player.prototype.destruction = function(place, length){
	var card = this.field.destruction(place, length);
	if(card.number == 0 || card.number == 1 || (card.number >= 11 && card.number <= 13)){
		this.alternation();
	}
	this.cemetery.add(card);
};

Player.prototype.alternation = function(){
	var card;
	const True = true;

	while(True){
		card = this.deck.getTop();
		if((card.number == 0 || card.number == 1 || (card.number >= 11 && card.number <= 13))){
			break;
		}
		this.cemetery.add(card);
	}
	this.hand.add(card);
};

Player.prototype.declarationDefense = function(soldierLength){
	var num = 0;
	for(var i = 0; i < soldierLength.length; i++){
		this.field.drive("soldier", soldierLength[i]);
		num += this.field.soldier[soldierLength[i]].number;
	}
	return num;
};

Player.prototype.declarationDefenseBarrier = function(length){
	this.field.drive("barrier", length);
	this.field.open(length);
};

Player.prototype.noCostSpell = function(handLength){
	this.cemetery.add(this.hand.get(handLength));
};

Player.prototype.onceSpell = function(handLength, costLength){
	this.cemetery.add(this.hand.get(costLength));
	this.cemetery.add(this.hand.get(handLength));
};

Player.prototype.twiceSpell = function(handLengths){
	this.cemetery.add(this.hand.get(handLengths[0]));
	this.cemetery.add(this.hand.get(handLengths[1]));
};

Player.prototype.up = function(handLength, costLength, length){
	var num = this.hand.cardList[handLength].number;
	this.field.soldier[length].attack += num;
	this.onceSpell(handLength, costLength);
};

Player.prototype.down = function(length, handLength, costLength){
	var num = this.hand.cardList[handLength].number;
	this.onceSpell(handLength, costLength);
	return num;
};

Player.prototype.twist = function(handLength, costLength, player, length){
	if(player){
		var num = this.hand.cardList[handLength].number;
		this.field.soldier[length].attack += num;
	}
	this.onceSpell(handLength, costLength);
};

Player.prototype.counter = function(handLength, costLength){
	this.onceSpell(handLength, costLength);
};

Player.prototype.barrierDefense = function(handLengths, player, length){
	if(player){
		this.destruction("barrier", length);
	}
	this.twiceSpell(handLengths);
};

Player.prototype.throwing = function(handLengths){
	this.twiceSpell(handLengths);
	if(handLengths[0].mark == "spade"){
		return handLengths[0].number;
	}else{
		return handLengths[1].number;
	}
};

Player.prototype.search = function(length, number, mark){
	this.noCostSpell(length);
	this.deck.search(number, mark);
};

module.exports = Player;

File: Python_Server/bp/Player.py
import Hand
import Field
import Deck
import Cemetery


class Player:

    """
    カードの召喚やドローなどプレイヤーの行動を定義

    Attribute
    ----------
    hand : Hand
        プレイヤーの手札
    field : Field
        プレイヤーのフィールド
    cemetery : Cemetery
        プレイヤーの墓地
    deck : Deck
        プレイヤーのデッキ
    can_attack
        プレイヤーが攻撃できるかどうか
    can_draw
        ドローできる限界
    """

    def __init__(self, deck_list: list):
        """
        コンストラクタ
        :param deck_list: デッキのリスト
        """

        self.hand = Hand.Hand()
        self.field = Field.Field()
        self.cemetery = Cemetery.Cemetery()
        self.deck = Deck.Deck(deck_list)
        self.deck.shuffle()
        self.can_attack = True
        self.can_draw = 7
        for i in range(0, 7):
            self.draw()

    # 一枚ドローして手札に加える
    def draw(self):
        self.hand.add(self.deck.draw())

    # numの数だけドローして手札に加える
    def draws(self, num: int):
        for i in range(0, num):
            self.hand.add(self.deck.draw())

    # ゲームの始めにデッキの一番上を一枚墓地に置く(先攻・後攻決め)
    def first_step(self) ->list:
        card = self.deck.first_step()
        self.cemetery.add(card)
        return [card.number, card.mark]

    # 自分のフィールドの兵士、防壁を全てチャージ状態にする
    def charge(self):
        self.field.charge_all()
        self.field.can_attack_all()

    # 手札が6枚を超えると手札の最後を墓地に置く
    def clean_up(self, length: int):
        if len(self.hand) > 7:
            self.cemetery.add(self.hand.get(length))

    # ダメージを受ける(デッキからダメージ文カードを墓地へ置く)
    def damage(self):
        self.cemetery.add(self.deck.damage())

    # 防壁の召喚
    def set_barrier(self, length: int):
        self.damage()
        self.field.set_barrier(self.hand.get(length))

    # 兵士の召喚
    def summon_soldier(self, hand_length: int, barrier_length: int):
        self.damage()
        self.field.drive("barrier", barrier_length)
        self.field.summon(self.hand.get(hand_length))

    # エースの召喚
    def summon_ace(self, hand_length: int):
        self.damage()
        self.field.summon(self.hand.get(hand_length))

    # 英雄の召喚
    def summon_hero(self, hand_length: int, barrier_length1: int, barrier_length2: int):
        self.damage()
        self.field.drive("barrier", barrier_length1)
        self.field.drive("barrier", barrier_length2)
        self.field.summon(self.hand.get(hand_length))

    # 魔術師の召喚
    def summon_magician(self, hand_length: int, cost_hand_length: int, barrier_length: int):
        self.field.drive("barrier", barrier_length)
        if hand_length > cost_hand_length:
            self.field.summon(self.hand.get(hand_length))
            self.cemetery.add(self.hand.get(cost_hand_length))
        else:
            self.cemetery.add(self.hand.get(cost_hand_length))
            self.field.summon(self.hand.get(hand_length))

    # 攻撃する兵士のリストを返す
    def declaration_attack(self, soldier_length: list) ->list:
        attacker = []
        for i in range(0, len(soldier_length)):
            if self.field.soldier[soldier_length[i]].can_attack:
                attacker.append(self.field.soldier[soldier_length[i]])
                self.field.drive("soldier", soldier_length[i])

        return attacker

    # 世代交代
    def alternation(self):
        while True:
            card = self.deck.get()
            if card.number == 0 or card.number == 1 or (11 <= card.number <= 13):
                break

            self.cemetery.add(card)
        self.hand.add(card)

    # フィールドからカードを取り除く
    def destruction(self, place: str, length: int):
        card = self.field.destruction(place, length)
        if card.number == 0 or card.number == 1 or (11 <= card.number <= 13):
            self.alternation()

        self.cemetery.add(card)

    # 兵士でブロック(複数可)
    def declaration_defense(self, soldier_length: list) ->int:
        num = 0
        for i in range(0, len(soldier_length)):
            self.field.drive("soldier", soldier_length[i])
            num += self.field.soldier[soldier_length[i]].number

        return num

    # 防壁でブロック
    def declaration_defense_barrier(self, length: int):
        self.field.drive("barrier", length)
        self.field.open(length)

    """未実装"""
    def no_cost_spell(self, hand_length: int):
        self.cemetery.add(self.hand.get(hand_length))

    def once_spell(self, hand_length: int, cost_length: int):
        self.cemetery.add(self.hand.get(cost_length))
        self.cemetery.add(self.hand.get(hand_length))

    def twice_spell(self, hand_lengths: int):
        self.cemetery.add(self.hand.get(hand_lengths[0]))
        self.cemetery.add(self.hand.get(hand_lengths[1]))

    def up(self, hand_length: int, cost_length: int, length: int):
        num = self.hand.card_list[hand_length].number
        self.field.soldier[length].attack += num
        self.once_spell(hand_length, cost_length)

    def down(self, hand_length: int, cost_length: int) ->int:
        num = self.hand.card_list[hand_length].number
        self.once_spell(hand_length, cost_length)
        return num

    def twist(self, hand_length: int, cost_length: int, player: bool, length: int):
        if player:
            num = self.hand.card_list[hand_length].number
            self.field.soldier[length].attack += num

        self.once_spell(hand_length, cost_length)

    def counter(self, hand_length, cost_length):
        self.once_spell(hand_length, cost_length)

    def barrier_defense(self, hand_length: int, player: bool, length: int):
        if player:
            self.destruction("barrier", length)

        self.twice_spell(hand_length)

    def throwing(self, hand_lengths: list) ->int:
        self.twice_spell(hand_lengths)
        if hand_lengths[0] == "spade":
            return hand_lengths[0].number
        else:
            return hand_lengths[1].number

    def search(self, length: int, number: int, mark: str):
        self.no_cost_spell(length)
        self.deck.search(number, mark)
